torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -579,3 +579,50 @@ TORCH_LIBRARY(_c10d_functional_autograd, m) {
           c10::DispatchKey::Autograd, ::all_gather_into_tensor_autograd),
       {at::Tag::pt2_compliant_tag});
 }
+
+namespace {
+// DTensor related comm operations, sharing code with functional collective for
+// now
+at::Tensor shard_dim_alltoall(
+    const at::Tensor& input,
+    int64_t gather_dim,
+    int64_t shard_dim,
+    std::string group_name) {
+  auto group = c10d::resolve_process_group(group_name);
+  auto group_size = group->getSize();
+  std::vector<int64_t> output_sizes = input.sizes().vec();
+  if (output_sizes[shard_dim] % group_size != 0) {
+    LOG(WARNING) << "The first dimension of the shard_dim_alltoall input ("
+                 << output_sizes[shard_dim]
+                 << ") is not divisible by the group size (" << group_size
+                 << ").";
+  }
+  output_sizes[shard_dim] = output_sizes[shard_dim] / group_size;
+  std::vector<at::Tensor> inputs;
+  auto length = output_sizes[shard_dim];
+  for (int i = 0; i < group_size; i++) {
+    inputs.push_back(input.narrow(shard_dim, i * length, length).contiguous());
+  }
+  // allocate outputs
+  std::vector<at::Tensor> outputs;
+  for (int i = 0; i < group_size; i++) {
+    outputs.push_back(input.new_empty(output_sizes).contiguous());
+  }
+  auto work = group->alltoall(outputs, inputs);
+
+  work->wait();
+  // TODO: it's very tricky to get the current async behavior work for shard dim
```

Comment:
We can certainly add this. Curious how desirable it is in this case?

Reply:
Currently this op is always sync, but if we can enable in async that would potentially help the case of eager mode reordering. I couldn't figure out how to do this cleanly yet so would appreciate your help here!

Essentially if the `gather_dim` is not 0, then we would need to torch.cat on the `outputs` on the `gather_dim`, but we cannot simply associate the `cat` output with the async work handle, as the `torch.cat` result is a new buffer than the output list

Reply:
Sounds good, I'll give it a try